### PR TITLE
[Constraint solver] De-prioritize SIMD operators. 

### PR DESCRIPTION
**Pull request by DougGregor**
The new SIMD proposal introduced a number of new operators, the presence of
which causes more "expression too complex" failures. Route around the
problem by de-prioritizing those operators, visiting them only if no
other operator could be chosen. This should limit the type checker
performance cost of said operators to only those expressions that need
them OR that already failed to type-check.

Fixes rdar://problem/46541800.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

File: lib/Sema/CSStep.cpp
Comment (xedin) [inline lib/Sema/CSStep.cpp]: Nit: can combine two ifs into one `!(func && function->isOperator())`, and same could be done for nominal using De Morgan's Law :)

File: lib/Sema/ConstraintSystem.cpp
Comment (xedin) [inline lib/Sema/ConstraintSystem.cpp]: I'm not sure if it's actually worse to check if we have any simd operators and then sort, might be faster to simplify sort since there are not going to be thousands of operators anyway...

**Review by xedin — APPROVED** [review]
LGTM! Unfortunately this means if `simd` is included and used it would be slower because we'd check generic operators and other stuff first...

File: lib/Sema/CSStep.cpp
Comment (DougGregor) [reply lib/Sema/CSStep.cpp]: I like breaking out the conditions here.

File: lib/Sema/ConstraintSystem.cpp
Comment (DougGregor) [reply lib/Sema/ConstraintSystem.cpp]: I'm trying to avoid the malloc + copy for the common case.

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

File: lib/Sema/ConstraintSystem.cpp
Comment (xedin) [reply lib/Sema/ConstraintSystem.cpp]: Thanks for clarification!

**Review by stephentyrone — APPROVED** [review]
SIMDOperators changes LGTM.

File: lib/Sema/ConstraintSystem.cpp
Comment (xedin) [reply lib/Sema/ConstraintSystem.cpp]: You can also use `llvm::find_if` to make it a bit smaller :)

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
Previous failure was a spurious one from LLDB:

```
TIMEOUT: test_dsym (lang/swift/variables/uninitialized/TestSwiftUninitializedVariable.py)
```

**Comment by DougGregor** [comment]
https://github.com/apple/swift/pull/21253 has a more advanced form of this hack, which should handle the newer test case @xedin came up with.

**Comment by DougGregor** [comment]
This optimization is safe and improves things a bit, so I'm going to go ahead and merge it. The more advanced hack in https://github.com/apple/swift/pull/21253 is also way more dicey, so I want to hold off on that.

**Pull request merged**
